sdks/python/apache_beam/pipeline.py

```diff
@@ -502,6 +505,8 @@ def run(self, test_runner_api='AUTO'):
     # Records whether this pipeline contains any cross-language transforms.
     self.contains_external_transforms = (
         ExternalTransformFinder.contains_external_transforms(self))
+    # Move in to_runner_api instead?
```

Comment:
I think it's fine to have here. (Preferable would be to actually build the things with the right hints in the first place rather than have a visitor modify them, but that's probably quite a bit more invasive of a change to take on now.)

Reply:
I think https://github.com/apache/beam/pull/14390/commits/5d58030998e77c066881891b94e4056c4778122a would take care of it? But if you think it's too risky, I can punt this commit for now.

Reply:
we may do multiple traversals though, I am actually not sure if that's better..

Reply:
Ah, the pipeline DAG of AppliedTransforms is constructed from root to leaves right? Then multiple traversals shouldn't be a concern, since we traverse at the time of application/adding a part.